### ISSUE TITLE
Linter to recommend using vectorized functions, when relevant?

### ISSUE THREAD

**Issue by IndrajeetPatil** [r-lib/lintr#2590]
Should there be a linter that catches usage of functionals (or simple for loops) with functions which are already vectorized? This is one of the most common mistakes programmers coming from other languages make: not utilizing the full potential of vectorization in R.

Of course, we can’t cover every function out there that is vectorized so the scope of this will have to be restricted to base-R and tidyverse packages (with argument to extend it somehow).

That said, I am not sure how we can determine if a function is vectorized. The simplest solution is of course to hardcode a list of known vectorized functions, but that will have to be a long list and will also be difficult to maintain. 

What do you think?

**Comment by MichaelChirico** [comment]
we have inner_combine_linter() and outer_negation_linter(), which are specific cases of this. unnecessary_lambda_linter() catches cases like sapply(x, \(xi) xi==2) also.

it's something good to improve but hard to detect statically, I'd go after incremental improvements if we see new rules similar to the above, I'm skeptical a fully general approach will bear fruit.

**Comment by IndrajeetPatil** [comment]
Yeah, I see what you mean. But it feels like an intimidating task to keep adding linters for every new case we find where such a linter might be relevant.

E.g. should there be a new linter to cover the case below? Or should we try to absorb such cases into (however loosely relevant) existing linter (say, `lengths_linter()`), and create a new linter only if we can't find any existing linter to be relevant?

``` r
library(stringr)

purrr::map_int(c("i", "like", "programming", NA), str_length)
#> [1]  1  4 11 NA

str_length(c("i", "like", "programming", NA))
#> [1]  1  4 11 NA
```

<sup>Created on 2024-06-04 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by AshesITR** [comment]
I wouldn't put this as beyond the scope of `unnecessary_lambda_linter()`.
Maybe a first step could be to extend this linter to also check for a whitelist of known-vectorized functions?

NB we could keep the whitelist empty or check which base R functions would cause a lot of lints on CRAN (= are useful to check) to keep our bookkeeping work minimal.
